src/test/ui/return/return-impl-trait.rs

```diff
@@ -0,0 +1,25 @@
+// run-rustfix
+
+trait Trait {}
+impl Trait for () {}
+
+// this works
+fn foo() -> impl Trait {
+    ()
+}
+
+fn bar<T: Trait + std::marker::Sync>() -> T
+//~^ this type parameter
+//~| expected `T` because of return type
+where
```

Comment:
So...what's difficult about this, is that you can have clauses like `Option<T>: Send`. This would effectively constrain `T: Send`, but wouldn't be picked up if T in the output is converted to `impl Trait`.

I think the above should be added as a test first and foremost. But, we shouldn't suggest changing if `T` appears in any clauses where it isn't the self type.